File: src/main/kotlin/tracking/website/TwitterProfileStrategy.kt
package tracking.website

import data.JsonException
import data.JsonNavigator
import io.github.oshai.kotlinlogging.KotlinLogging
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import web.url.FreeUrl
import web.url.Url

private val logger = KotlinLogging.logger {}

object TwitterProfileStrategy : Strategy {
    private val cookieInitUrl = FreeUrl("https://twitter.com/", StandardStrategy)
    private val loginLocationHeader = Regex("^location: \\S+login\\S+$", setOf(RegexOption.MULTILINE, RegexOption.IGNORE_CASE))
    private val profileUrlRegex = Regex("https?://(m\\.|www\\.)?twitter\\.com/[^/?]+(\\?.*)?")

    override fun isSuitableFor(url: String) = profileUrlRegex.matches(url)

    override fun filterContents(input: String): String {
        val document = Jsoup.parse(input)

        return tryFromLdJsonProfilePage(document) ?: tryFromMeta(document) ?: input
    }

    private fun tryFromMeta(document: Document): String? {
        val ogDescriptionNodes = document.head().selectXpath("//meta[@property='og:description']") // grep-code-og-description-removal
        val ogTitleNodes = document.head().selectXpath("//meta[@property='og:title']")

        val ogTitle: String? = ogTitleNodes.attr("content")
        val ogDescription: String? = ogDescriptionNodes.attr("content")

        if (ogTitle == null || ogDescription == null) {
            logger.warn { "Failed reading og:title and/or og:description nodes" }

            return null
        }

        return ogTitle + "\n" + ogDescription
    }

    private fun tryFromLdJsonProfilePage(document: Document): String? {
        val ldJsonNodes = document
            .head()
            .selectXpath("//script[@type='application/ld+json'][contains(text(), 'ProfilePage')]")

        if (ldJsonNodes.isEmpty()) {
            logger.info { "Failed to XPath Twitter ProfilePage schema script node" }

            return null
        }

        return try {
            val ldJsonData = JsonNavigator(ldJsonNodes.html())

            ldJsonData.getString("author/givenName") +
                    "\n" + ldJsonData.getString("author/description") +
                    "\n" + ldJsonData.getString("author/homeLocation/name")
        } catch (exception: JsonException) {
            logger.warn(exception) { "Failed reading ProfilePage schema JSON" }

            null
        }
    }

    override fun getLatentCode(url: Url, contents: String, originalCode: Int): Int {
        return if (originalCode == 302 && contents.contains(loginLocationHeader)) {
            401 // SPACE KAREN
        } else {
            originalCode
        }
    }

    override fun getCookieInitUrl() = cookieInitUrl
}

File: symfony/src/Utils/Web/UrlStrategy/TwitterProfileStrategy.php
<?php

declare(strict_types=1);

namespace App\Utils\Web\UrlStrategy;

use App\Utils\Web\Url\FreeUrl;
use App\Utils\Web\Url\Url;
use App\Utils\Web\Url\UrlForTracking;
use Composer\Pcre\Preg;
use Override;
use Symfony\Component\DomCrawler\Crawler;

/**
 * As of 2024-06-14, Twitter's behavior is to send around 3700 bytes of the page,
 * often enough to cover the description and title.
 */
class TwitterProfileStrategy extends Strategy
{
    private const string profileUrlRegex = '~^https://(twitter|x)\.com/(?<username>[^/?]+)$~';

    #[Override]
    public static function isSuitableFor(string $url): bool
    {
        return Preg::isMatch(self::profileUrlRegex, $url);
    }

    #[Override]
    public function getUrlForTracking(Url $url): Url
    {
        return new UrlForTracking(str_replace('https://twitter.com/', 'https://x.com/', $url->getUrl()), $url); // Space Karen
    }

    #[Override]
    public function filterContents(string $input): string
    {
        $crawler = new Crawler($input);

        $ogDescriptionNodes = $crawler->filterXPath("//head/meta[@property='og:description']");
        $ogTitleNodes = $crawler->filterXPath("//head/meta[@property='og:title']");

        if (0 === $ogDescriptionNodes->count() || 0 === $ogTitleNodes->count()) {
            return $input;
        }

        $ogTitle = $ogTitleNodes->attr('content') ?? '';
        $ogDescription = $ogDescriptionNodes->attr('content') ?? '';

        return "$ogTitle\n$ogDescription";
    }

    public function getCookieInitUrl(): Url
    {
        return new FreeUrl('https://x.com/', ''); // Space Karen
    }
}
